docs/application-settings.rst
.. _scancodeio_settings:

Application Settings
====================

ScanCode.io is configured with environment variables stored in a ``.env`` file.

The ``.env`` file is created at the root of the ScanCode.io codebase during its
installation.
You can configure your preferences using the following settings in the ``.env``
file.

.. note::
    ScanCode.io is based on the Django web framework and its settings system.
    The list of settings available in Django is documented at
    `Django Settings <https://docs.djangoproject.com/en/dev/ref/settings/>`_.

.. tip::
    Settings specific to ScanCode.io are all prefixed with ``SCANCODEIO_``.

**Restarting the services is required following any changes to .env:**

.. code-block:: bash

    docker compose restart web worker

Instance settings
-----------------

DATABASE
^^^^^^^^

The database can be configured using the following settings::

    SCANCODEIO_DB_HOST=localhost
    SCANCODEIO_DB_NAME=scancodeio
    SCANCODEIO_DB_USER=user
    SCANCODEIO_DB_PASSWORD=password
    SCANCODEIO_DB_PORT=5432

.. _scancodeio_settings_require_authentication:

SCANCODEIO_REQUIRE_AUTHENTICATION
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

By default, the ScanCode.io Web UI and REST API are available without any
authentication.

The authentication system can be enable with this settings::

    SCANCODEIO_REQUIRE_AUTHENTICATION=True

Once enabled, all the Web UI views and REST API endpoints will force the user to login
to gain access.

A management command :ref:`cli_create_user` is available to create users and
generate their API key for authentication.

See :ref:`rest_api_authentication` for details on using the ``API key``
authentication system in the REST API.

.. _scancodeio_settings_workspace_location:

SCANCODEIO_WORKSPACE_LOCATION
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This setting defines the workspace location of a given project.
The **workspace** is the directory where **all of the project's files are stored**
, such as input, codebase, and output files::

    SCANCODEIO_WORKSPACE_LOCATION=/var/scancodeio/workspace/

It defaults to a :guilabel:`var/` directory in the local ScanCode.io codebase.

See :ref:`project_workspace` for more details.

.. _scancodeio_settings_config_dir:

SCANCODEIO_CONFIG_DIR
^^^^^^^^^^^^^^^^^^^^^

The location of the :guilabel:`.scancode/` configuration directory within the project
codebase.

Default: ``.scancode``

This directory allows to provide configuration files and customization for a ScanCode.io
project directly through the codebase files.

For example, to provide a custom attribution template to your project, add it in a
:guilabel:`.scancode/` directory located at the root of your codebase before uploading
it to ScanCode.io. The expected location of the attribution template is::

  .scancode/templates/attribution.html

SCANCODEIO_PROCESSES
^^^^^^^^^^^^^^^^^^^^

By default, multiprocessing is enabled and configured to use an optimal number of CPUs
available on the machine. You can control the number of parallel processes available
to ScanCode.io using the SCANCODEIO_PROCESSES setting::

    SCANCODEIO_PROCESSES=4

Multiprocessing can be disabled entirely using "0"::

    SCANCODEIO_PROCESSES=0

To disable both multiprocessing and threading, use "-1"::

    SCANCODEIO_PROCESSES=-1

.. note::
    Multiprocessing and threading are disabled by default on operating system
    where the multiprocessing start method is not "fork", such as on macOS.

.. _scancodeio_settings_async:

SCANCODEIO_ASYNC
^^^^^^^^^^^^^^^^

When enabled, pipeline runs are **executed asynchronously**, meaning that users can
continue using the app while the pipeline are run in the background.

The ASYNC mode is **enabled by default in a "Run with Docker" configuration** but
**disabled in a "Local development" setup**.

It is possible to enable ASYNC mode in a "local development" setup with the following
setting::

    SCANCODEIO_ASYNC=True

Once enabled, pipeline runs will be sent to a task queue instead of being executed
synchronously in the web server process.

.. warning::
    The ASYNC mode required a **Redis server** and running a **tasks worker** using
    ``$ make worker``.

    On macOS, the ASYNC mode requires the following line in your environment::

        export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES

SCANCODEIO_TASK_TIMEOUT
^^^^^^^^^^^^^^^^^^^^^^^

Maximum time allowed for a pipeline to complete.
The pipeline run will be stopped and marked as failed if that limit is reached.

The value is a string with specify unit including hour, minute, second
(e.g. "1h", "3m", "5s")::

    SCANCODEIO_TASK_TIMEOUT=24h

Default: ``24h``

SCANCODEIO_SCAN_FILE_TIMEOUT
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Maximum time allowed for a file to be analyzed when scanning a codebase.

The value unit is second and is defined as an integer::

    SCANCODEIO_SCAN_FILE_TIMEOUT=120

Default: ``120`` (2 minutes)

.. _scancodeio_settings_pipelines_dirs:

SCANCODEIO_PIPELINES_DIRS
^^^^^^^^^^^^^^^^^^^^^^^^^

This setting defines any additional locations that ScanCode.io will search in
for pipelines.
It usually includes a list of comma-separated strings containing full paths
of additional pipelines directories::

    SCANCODEIO_PIPELINES_DIRS=/var/scancodeio/pipelines/,/home/user/pipelines/

.. _scancodeio_settings_policies_file:

SCANCODEIO_POLICIES_FILE
^^^^^^^^^^^^^^^^^^^^^^^^

This setting defines the location of the policies file, or ``policies.yml``.
A valid policies file is required to enable compliance-related features.

.. code-block:: yaml

    license_policies:
    -   license_key: mit
        label: Approved License
        compliance_alert: ''
    -   license_key: mpl-2.0
        label: Restricted License
        compliance_alert: warning
    -   license_key: gpl-3.0
        label: Prohibited License
        compliance_alert: error

- Licenses are referenced by the ``license_key``.
- A Policy is defined with ``label`` and ``compliance_alert``.
- The ``compliance_alert`` accepts 3 values: '' for an empty string, warning, and error.

.. note::
    When the policy feature is enabled, the ``compliance_alert`` values are
    displayed in the UI and returned in all downloadable results.

.. tip::
    Check out the :ref:`tutorial_license_policies` tutorial for in-depth coverage of
    this feature.

SCANCODEIO_PAGINATE_BY
^^^^^^^^^^^^^^^^^^^^^^

The number of objects display per page for each object type can be customized with the
following setting::

    SCANCODEIO_PAGINATE_BY=project=30,error=50,resource=100,package=100,dependency=100

SCANCODEIO_REST_API_PAGE_SIZE
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A numeric value indicating the number of objects returned per page in the REST API::

    SCANCODEIO_REST_API_PAGE_SIZE=100

Default: ``50``

.. warning::
    Using a large page size may have an impact on performances.

SCANCODEIO_LOG_LEVEL
^^^^^^^^^^^^^^^^^^^^

By default, only a minimum of logging messages is displayed in the console, mostly
to provide some progress about pipeline run execution.

Default: ``INFO``

The ``DEBUG`` value can be provided to this setting to see all ScanCode.io debug
messages to help track down configuration issues for example.
This mode can be enabled globally through the ``.env`` file::

    SCANCODEIO_LOG_LEVEL=DEBUG

Or, in the context of running a :ref:`scanpipe command <command_line_interface>`:

.. code-block:: console

    $ SCANCODEIO_LOG_LEVEL=DEBUG bin/scanpipe [command]

The web server can be started in DEBUG mode with:

.. code-block:: console

    $ SCANCODEIO_LOG_LEVEL=DEBUG make run

TIME_ZONE
^^^^^^^^^

A string representing the time zone for the current ScanCode.io installation. By
default the ``UTC`` time zone is used::

    TIME_ZONE=Europe/Paris

.. note::
    You can view a detailed list of time zones `here.
    <https://en.wikipedia.org/wiki/List_of_tz_database_time_zones>`_

.. _scancodeio_settings_external_services:

External services (integrations)
--------------------------------

.. _scancodeio_settings_purldb:

PURLDB
^^^^^^

A public instance of **PurlDB** is accessible at https://public.purldb.io/.

Alternatively, you can deploy your own instance of PurlDB by
following the instructions provided in the documentation at
https://purldb.readthedocs.io/.

To configure your local environment, set the ``PURLDB_URL`` in your ``.env`` file::

    PURLDB_URL=https://public.purldb.io/

While using the public PurlDB instance, providing an API key is optional.
However, if authentication is enabled on your PurlDB instance, you can provide the
API key using ``PURLDB_API_KEY``::

    PURLDB_API_KEY=insert_your_api_key_here

.. note::
    Once the PurlDB is configured, a new "PurlDB" tab will be available in the
    discovered package details view.

.. _scancodeio_settings_vulnerablecode:

VULNERABLECODE
^^^^^^^^^^^^^^

You have the option to either deploy your instance of
`VulnerableCode <https://github.com/aboutcode-org/vulnerablecode/>`_
or connect to the `public instance <https://public.vulnerablecode.io/>`_.

To configure your local environment, set the ``VULNERABLECODE_URL`` in your ``.env``
file::

    VULNERABLECODE_URL=https://public.vulnerablecode.io/

When using the public VulnerableCode instance, providing an API key is optional.
However, if authentication is enabled on your VulnerableCode instance,
you can provide the API key using ``VULNERABLECODE_API_KEY``::

    VULNERABLECODE_API_KEY=insert_your_api_key_here

.. _scancodeio_settings_matchcodeio:

MATCHCODE.IO
^^^^^^^^^^^^

There is currently no public instance of MatchCode.io.

Alternatively, you can deploy your own instance of MatchCode.io by
following the instructions provided in the documentation at
https://purldb.readthedocs.io/.

To configure your local environment, set the ``MATCHCODEIO_URL`` in your ``.env`` file::

    MATCHCODEIO_URL=https://<Address to MatchCode.io host>/

If authentication is enabled on your MatchCode.io instance, you can provide the
API key using ``MATCHCODEIO_API_KEY``::

    MATCHCODEIO_API_KEY=insert_your_api_key_here

.. _scancodeio_settings_fetch_authentication:

Fetch Authentication
--------------------

Several settings are available to define the credentials required to access your
private files, depending on the authentication type: Basic, Digest, Token header, etc.

.. note::
    The provided credentials are enabled for all projects on the ScanCode.io instance.

.. warning::
    Ensure that the provided ``host`` values are fully qualified, including the domain
    and subdomain.

.. _scancodeio_settings_fetch_basic_auth:

SCANCODEIO_FETCH_BASIC_AUTH
^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can provide credentials for input URLs protected by Basic Authentication using
the ``host=user,password`` syntax::

    SCANCODEIO_FETCH_BASIC_AUTH="www.host1.com=user,password;www.host2.com=user,password;"

.. _scancodeio_settings_fetch_digest_auth:

SCANCODEIO_FETCH_DIGEST_AUTH
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can provide credentials for input URLs protected by Digest Authentication using
the ``host=user,password`` syntax::

    SCANCODEIO_FETCH_DIGEST_AUTH="www.host1.com=user,password;www.host2.com=user,password;"

.. _scancodeio_settings_fetch_headers:

SCANCODEIO_FETCH_HEADERS
^^^^^^^^^^^^^^^^^^^^^^^^

When authentication credentials can be provided through HTTP request headers, you can
use the following syntax::

    SCANCODEIO_FETCH_HEADERS="www.host1.com=Header1=value,Header2=value;"

Example for a GitHub private repository::

    SCANCODEIO_FETCH_HEADERS="raw.github.com=Authorization=token <YOUR_TOKEN>"

.. _scancodeio_settings_netrc_location:

SCANCODEIO_NETRC_LOCATION
^^^^^^^^^^^^^^^^^^^^^^^^^

If your credentials are stored in a
`.netrc <https://everything.curl.dev/usingcurl/netrc>`_ file, you can provide its
location on disk using::

    SCANCODEIO_NETRC_LOCATION="~/.netrc"

If you are deploying ScanCode.io using Docker and you wish to use a netrc file,
you can provide it to the Docker container by moving the netrc file to
``/etc/scancodeio/.netrc`` and then updating the ``.env`` file with the line::

    SCANCODEIO_NETRC_LOCATION="/etc/scancodeio/.netrc"

.. _scancodeio_settings_skopeo_credentials:

SCANCODEIO_SKOPEO_CREDENTIALS
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can define the username and password for Skopeo to access containers private
registries using the ``host=user:password`` syntax::

  SCANCODEIO_SKOPEO_CREDENTIALS="host1=user:password,host2=user:password"

.. _scancodeio_settings_skopeo_authfile_location:

SCANCODEIO_SKOPEO_AUTHFILE_LOCATION
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Specify the path of the Skopeo authentication file using the following setting::

    SCANCODEIO_SKOPEO_AUTHFILE_LOCATION="/path/to/auth.json"


scancodeio/settings.py
# SPDX-License-Identifier: Apache-2.0
#
# http://nexb.com and https://github.com/nexB/scancode.io
# The ScanCode.io software is licensed under the Apache License version 2.0.
# Data generated with ScanCode.io is provided as-is without warranties.
# ScanCode is a trademark of nexB Inc.
#
# You may not use this software except in compliance with the License.
# You may obtain a copy of the License at: http://apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.
#
# Data Generated with ScanCode.io is provided on an "AS IS" BASIS, WITHOUT WARRANTIES
# OR CONDITIONS OF ANY KIND, either express or implied. No content created from
# ScanCode.io should be considered or used as legal advice. Consult an Attorney
# for any legal advice.
#
# ScanCode.io is a free software code scanning tool from nexB Inc. and others.
# Visit https://github.com/nexB/scancode.io for support and download.

import sys
import tempfile
from pathlib import Path

import environ

PROJECT_DIR = environ.Path(__file__) - 1
ROOT_DIR = PROJECT_DIR - 1

# Environment

ENV_FILE = "/etc/scancodeio/.env"
if not Path(ENV_FILE).exists():
    ENV_FILE = ROOT_DIR(".env")

env = environ.Env()
environ.Env.read_env(ENV_FILE)

# Security

SECRET_KEY = env.str("SECRET_KEY")

ALLOWED_HOSTS = env.list(
    "ALLOWED_HOSTS",
    default=[".localhost", "127.0.0.1", "[::1]", "host.docker.internal", "172.17.0.1"],
)

CSRF_TRUSTED_ORIGINS = env.list("CSRF_TRUSTED_ORIGINS", default=[])

# SECURITY WARNING: don't run with debug turned on in production
DEBUG = env.bool("SCANCODEIO_DEBUG", default=False)

SCANCODEIO_REQUIRE_AUTHENTICATION = env.bool(
    "SCANCODEIO_REQUIRE_AUTHENTICATION", default=False
)

SCANCODEIO_ENABLE_ADMIN_SITE = env.bool("SCANCODEIO_ENABLE_ADMIN_SITE", default=False)

SECURE_CONTENT_TYPE_NOSNIFF = env.bool("SECURE_CONTENT_TYPE_NOSNIFF", default=True)

X_FRAME_OPTIONS = env.str("X_FRAME_OPTIONS", default="DENY")

SESSION_COOKIE_SECURE = env.bool("SESSION_COOKIE_SECURE", default=True)

CSRF_COOKIE_SECURE = env.bool("CSRF_COOKIE_SECURE", default=True)

# ``security.W004`` SECURE_HSTS_SECONDS and ``security.W008`` SECURE_SSL_REDIRECT
# are handled by the web server.
SILENCED_SYSTEM_CHECKS = ["security.W004", "security.W008"]

# ScanCode.io

SCANCODEIO_WORKSPACE_LOCATION = env.str("SCANCODEIO_WORKSPACE_LOCATION", default="var")

SCANCODEIO_CONFIG_DIR = env.str("SCANCODEIO_CONFIG_DIR", default=".scancode")

SCANCODEIO_CONFIG_FILE = env.str(
    "SCANCODEIO_CONFIG_FILE", default="scancode-config.yml"
)

SCANCODEIO_LOG_LEVEL = env.str("SCANCODEIO_LOG_LEVEL", "INFO")

# Set the number of parallel processes to use for ScanCode related scan execution.
# If the SCANCODEIO_PROCESSES argument is not set, defaults to an optimal number of CPUs
# available on the machine.
SCANCODEIO_PROCESSES = env.int("SCANCODEIO_PROCESSES", default=None)

SCANCODEIO_POLICIES_FILE = env.str("SCANCODEIO_POLICIES_FILE", default="policies.yml")

# This setting defines the additional locations ScanCode.io will search for pipelines.
# This should be set to a list of strings that contain full paths to your additional
# pipelines directories.
SCANCODEIO_PIPELINES_DIRS = env.list("SCANCODEIO_PIPELINES_DIRS", default=[])

# Maximum time allowed for a pipeline to complete.
SCANCODEIO_TASK_TIMEOUT = env.str("SCANCODEIO_TASK_TIMEOUT", default="24h")

# Default to 2 minutes.
SCANCODEIO_SCAN_FILE_TIMEOUT = env.int("SCANCODEIO_SCAN_FILE_TIMEOUT", default=120)

# List views pagination, controls the number of items displayed per page.
# Syntax in .env: SCANCODEIO_PAGINATE_BY=project=10,project_error=10
SCANCODEIO_PAGINATE_BY = env.dict(
    "SCANCODEIO_PAGINATE_BY",
    default={
        "project": 20,
        "error": 50,
        "resource": 100,
        "package": 100,
        "dependency": 100,
        "license": 100,
        "relation": 100,
    },
)

# Default limit for "most common" entries in QuerySets.
SCANCODEIO_MOST_COMMON_LIMIT = env.int("SCANCODEIO_MOST_COMMON_LIMIT", default=7)

# Fetch authentication credentials

# SCANCODEIO_FETCH_BASIC_AUTH="host=user,password;"
SCANCODEIO_FETCH_BASIC_AUTH = env.dict(
    "SCANCODEIO_FETCH_BASIC_AUTH",
    cast={"value": tuple},
    default={},
)

# SCANCODEIO_FETCH_DIGEST_AUTH="host=user,password;"
SCANCODEIO_FETCH_DIGEST_AUTH = env.dict(
    "SCANCODEIO_FETCH_DIGEST_AUTH",
    cast={"value": tuple},
    default={},
)

# SCANCODEIO_FETCH_HEADERS="host=Header1=value,Header2=value;"
SCANCODEIO_FETCH_HEADERS = {}
FETCH_HEADERS_STR = env.str("SCANCODEIO_FETCH_HEADERS", default="")
for entry in FETCH_HEADERS_STR.split(";"):
    if entry.strip():
        host, headers = entry.split("=", 1)
        SCANCODEIO_FETCH_HEADERS[host] = env.parse_value(headers, cast=dict)

# SCANCODEIO_NETRC_LOCATION="~/.netrc"
SCANCODEIO_NETRC_LOCATION = env.str("SCANCODEIO_NETRC_LOCATION", default="")
if SCANCODEIO_NETRC_LOCATION:
    # Propagate the location to the environ for `requests.utils.get_netrc_auth`
    env.ENVIRON["NETRC"] = SCANCODEIO_NETRC_LOCATION

# SCANCODEIO_SKOPEO_CREDENTIALS="host1=user:password,host2=user:password"
SCANCODEIO_SKOPEO_CREDENTIALS = env.dict("SCANCODEIO_SKOPEO_CREDENTIALS", default={})

# SCANCODEIO_SKOPEO_AUTHFILE_LOCATION="/path/to/auth.json"
SCANCODEIO_SKOPEO_AUTHFILE_LOCATION = env.str(
    "SCANCODEIO_SKOPEO_AUTHFILE_LOCATION", default=""
)

# Application definition

INSTALLED_APPS = [
    # Local apps
    # Must come before Third-party apps for proper templates override
    "scanpipe",
    # Django built-in
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "django.contrib.admin",
    "django.contrib.humanize",
    # Third-party apps
    "crispy_forms",
    "crispy_bootstrap3",  # required for the djangorestframework browsable API
    "django_filters",
    "rest_framework",
    "rest_framework.authtoken",
    "django_rq",
    "django_probes",
    "fontawesomefree",
    "taggit",
]

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
]

ROOT_URLCONF = "scancodeio.urls"

WSGI_APPLICATION = "scancodeio.wsgi.application"

SECURE_PROXY_SSL_HEADER = env.tuple(
    "SECURE_PROXY_SSL_HEADER", default=("HTTP_X_FORWARDED_PROTO", "https")
)

# Database

DATABASES = {
    "default": {
        "ENGINE": env.str("SCANCODEIO_DB_ENGINE", "django.db.backends.postgresql"),
        "HOST": env.str("SCANCODEIO_DB_HOST", "localhost"),
        "NAME": env.str("SCANCODEIO_DB_NAME", "scancodeio"),
        "USER": env.str("SCANCODEIO_DB_USER", "scancodeio"),
        "PASSWORD": env.str("SCANCODEIO_DB_PASSWORD", "scancodeio"),
        "PORT": env.str("SCANCODEIO_DB_PORT", "5432"),
        "ATOMIC_REQUESTS": True,
    }
}

DEFAULT_AUTO_FIELD = "django.db.models.AutoField"

# Forms and filters

FILTERS_EMPTY_CHOICE_LABEL = env.str("FILTERS_EMPTY_CHOICE_LABEL", default="All")

# Templates

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "APP_DIRS": True,
        "OPTIONS": {
            "debug": DEBUG,
            "context_processors": [
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
                "django.template.context_processors.request",
                "scancodeio.context_processors.versions",
            ],
        },
    },
]

# Login

LOGIN_REDIRECT_URL = "project_list"

# Passwords

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
        "OPTIONS": {
            "min_length": env.int("SCANCODEIO_PASSWORD_MIN_LENGTH", default=12),
        },
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
]

# Testing

# True if running tests through `./manage test`
IS_TESTS = "test" in sys.argv

if IS_TESTS:
    # Do not pollute the workspace while running the tests.
    SCANCODEIO_WORKSPACE_LOCATION = tempfile.mkdtemp()
    SCANCODEIO_REQUIRE_AUTHENTICATION = True
    SCANCODEIO_SCAN_FILE_TIMEOUT = 120
    # The default password hasher is rather slow by design.
    # Using a faster hashing algorithm in the testing context to speed up the run.
    PASSWORD_HASHERS = [
        "django.contrib.auth.hashers.MD5PasswordHasher",
    ]

# Debug toolbar

DEBUG_TOOLBAR = env.bool("SCANCODEIO_DEBUG_TOOLBAR", default=False)
if DEBUG and DEBUG_TOOLBAR:
    INSTALLED_APPS.append("debug_toolbar")
    MIDDLEWARE.append("debug_toolbar.middleware.DebugToolbarMiddleware")
    INTERNAL_IPS = ["127.0.0.1"]

# Logging

LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "simple": {
            "format": "{levelname} {message}",
            "style": "{",
        },
    },
    "handlers": {
        "null": {
            "class": "logging.NullHandler",
        },
        "console": {
            "class": "logging.StreamHandler",
            "formatter": "simple",
        },
    },
    "loggers": {
        "scanpipe": {
            "handlers": ["null"] if IS_TESTS else ["console"],
            "level": SCANCODEIO_LOG_LEVEL,
            "propagate": False,
        },
        "django": {
            "handlers": ["null"] if IS_TESTS else ["console"],
            "propagate": False,
        },
        # Set SCANCODEIO_LOG_LEVEL=DEBUG to display all SQL queries in the console.
        "django.db.backends": {
            "level": SCANCODEIO_LOG_LEVEL,
        },
    },
}

# Instead of sending out real emails the console backend just writes the emails
# that would be sent to the standard output.
EMAIL_BACKEND = "django.core.mail.backends.console.EmailBackend"

# Internationalization

LANGUAGE_CODE = "en-us"

TIME_ZONE = env.str("TIME_ZONE", default="UTC")

USE_I18N = True

USE_TZ = True

# Static files (CSS, JavaScript, Images)

STATIC_URL = "/static/"

STATIC_ROOT = env.str("STATIC_ROOT", default="/var/scancodeio/static/")

STATICFILES_DIRS = [
    PROJECT_DIR("static"),
]

# Third-party apps

CRISPY_TEMPLATE_PACK = "bootstrap3"

# Job Queue

RQ_QUEUES = {
    "default": {
        "HOST": env.str("SCANCODEIO_REDIS_HOST", default="localhost"),
        "PORT": env.str("SCANCODEIO_REDIS_PORT", default="6379"),
        "PASSWORD": env.str("SCANCODEIO_REDIS_PASSWORD", default=""),
        "DEFAULT_TIMEOUT": env.int("SCANCODEIO_REDIS_DEFAULT_TIMEOUT", default=360),
    },
}

SCANCODEIO_ASYNC = env.bool("SCANCODEIO_ASYNC", default=False)
if not SCANCODEIO_ASYNC:
    for queue_config in RQ_QUEUES.values():
        queue_config["ASYNC"] = False

# ClamAV virus scan
CLAMD_USE_TCP = env.bool("CLAMD_USE_TCP", default=True)
CLAMD_TCP_ADDR = env.str("CLAMD_TCP_ADDR", default="clamav")

# Django restframework

REST_FRAMEWORK = {
    "DEFAULT_AUTHENTICATION_CLASSES": (
        "rest_framework.authentication.TokenAuthentication",
    ),
    "DEFAULT_PERMISSION_CLASSES": ("rest_framework.permissions.IsAuthenticated",),
    "DEFAULT_RENDERER_CLASSES": (
        "rest_framework.renderers.JSONRenderer",
        "rest_framework.renderers.BrowsableAPIRenderer",
        "rest_framework.renderers.AdminRenderer",
    ),
    "DEFAULT_FILTER_BACKENDS": (
        "django_filters.rest_framework.DjangoFilterBackend",
        "rest_framework.filters.SearchFilter",
    ),
    "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.PageNumberPagination",
    "PAGE_SIZE": env.int("SCANCODEIO_REST_API_PAGE_SIZE", default=50),
    "UPLOADED_FILES_USE_URL": False,
}

if not SCANCODEIO_REQUIRE_AUTHENTICATION:
    REST_FRAMEWORK["DEFAULT_PERMISSION_CLASSES"] = (
        "rest_framework.permissions.AllowAny",
    )

# VulnerableCode integration

VULNERABLECODE_URL = env.str("VULNERABLECODE_URL", default="")
VULNERABLECODE_USER = env.str("VULNERABLECODE_USER", default="")
VULNERABLECODE_PASSWORD = env.str("VULNERABLECODE_PASSWORD", default="")
VULNERABLECODE_API_KEY = env.str("VULNERABLECODE_API_KEY", default="")

# PurlDB integration

PURLDB_URL = env.str("PURLDB_URL", default="")
PURLDB_USER = env.str("PURLDB_USER", default="")
PURLDB_PASSWORD = env.str("PURLDB_PASSWORD", default="")
PURLDB_API_KEY = env.str("PURLDB_API_KEY", default="")

# MatchCode.io integration

MATCHCODEIO_URL = env.str("MATCHCODEIO_URL", default="")
MATCHCODEIO_USER = env.str("MATCHCODEIO_USER", default="")
MATCHCODEIO_PASSWORD = env.str("MATCHCODEIO_PASSWORD", default="")
MATCHCODEIO_API_KEY = env.str("MATCHCODEIO_API_KEY", default="")
